src/backend/distributed/planner/recursive_planning.c

```diff
@@ -316,6 +302,211 @@ RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
+ */
+static bool
+RecursivelyPlanSubqueryWalker(Node *node, RecursivePlanningContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		DeferredErrorMessage *error = NULL;
+
+		context->level += 1;
+		error = RecursivelyPlanSubqueriesAndCTEs(query, context);
+		if (error != NULL)
+		{
+			RaiseDeferredError(error, ERROR);
+		}
+		context->level -= 1;
+
+		if (ShouldRecursivelyPlanSubquery(query, context))
+		{
+			RecursivelyPlanSubquery(query, context);
+		}
+
+		return false;
+	}
+
+	return expression_tree_walker(node, RecursivelyPlanSubqueryWalker, context);
+}
+
+
+/*
+ * ShouldRecursivelyPlanSubquery decides whether the input subquery should be recursively
+ * planned or not.
+ *
+ * For the details, see the cases in the function.
+ */
+static bool
+ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context)
+{
+	bool shouldRecursivelyPlan = false;
+
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		/*
+		 * Postgres can always plan queries that don't require distributed planning.
+		 * Note that we need to check this first, otherwise the calls to the many other
+		 * Citus planner functions would error our due to local relations.
+		 */
+		shouldRecursivelyPlan = true;
+	}
+	else
+	{
+		if (DeferErrorIfCannotPushdownSubquery(subquery, false) == NULL)
+		{
+			/*
+			 * Citus can pushdown this subquery, no need to recursively
+			 * plan which is much expensive than pushdown.
+			 */
+			shouldRecursivelyPlan = false;
+		}
+		else if (TaskExecutorType == MULTI_EXECUTOR_TASK_TRACKER &&
+				 SingleRelationRepartitionSubquery(subquery))
+		{
+			/* Citus could plan this subquery through re-partitioning */
+			shouldRecursivelyPlan = false;
+		}
+		else if (DeferErrorIfQueryNotSupported(subquery) == NULL)
+		{
+			/*
+			 * At this point, we might be passing a query with or without
+			 * subqueries to DeferErrorIfQueryNotSupported().
+			 *
+			 * In case of a query without any subqueries, we're OK given that
+			 * DeferErrorIfQueryNotSupported() is purely designed for such queries.
+			 *
+			 * In case of a query with subqueries in it, we're still OK given that
+			 * we had already checked those subqueries via the previous recursive calls
+			 * and each of the subqueries are guaranteed to be safe to pushdown or
+			 * already replaced with intermediate results. Thus, we're only interested
+			 * in the checks that we do do the top-level query.
+			 */
+			shouldRecursivelyPlan = true;
+		}
+		else
+		{
+			/*
+			 * XXX: What about queries that are router plannable but not
+			 * suppoted by logical planner such as windown functions.
+			 */
+		}
+	}
+
+	/*
+	 * Even if we could recursively plan the subquery, we should ensure
+	 * that the subquery doesn't contain any references to the outer
+	 * queries.
+	 */
+	if (shouldRecursivelyPlan && ContainsReferencesToOuterQuery(subquery))
```

Comment:
Maybe move this to the start of the if-else block above.

Reply:
Nvm, at the end probably makes sense given the debug message. Maybe split this into two functions in which only the outer function emits debug messages.